projects/hip-tests/catch/perftests/graph/hipGraphTopology.cc

```diff
@@ -0,0 +1,462 @@
+          current_batch_size++;
+        }
+      }
+      node_types.push_back(node_type);
+      printf("Node %3d: %-8s", i, node_type.c_str());
+      if (i % 9 == 8 || i == opt.length - 1) printf("\n");
+      else printf(" -> ");
+    }
+    
+    if (current_batch_size > 0) {
+      kernel_batches.push_back(current_batch_size);
+    }
+    
+    printf("\nMixed topology summary:\n");
+    printf("Total nodes: %d\n", opt.length);
+    printf("Kernel nodes: %d\n", kernel_count);
+    printf("Non-kernel nodes: %d\n", opt.length - kernel_count);
+    printf("Potential kernel batches: %d\n", (int)kernel_batches.size());
+    printf("Batch sizes: ");
+    for (size_t i = 0; i < kernel_batches.size(); ++i) {
+      printf("%d", kernel_batches[i]);
+      if (i < kernel_batches.size() - 1) printf(", ");
+    }
+    printf("\n\n");
+
+  } else if (opt.topology == "hexagon") {
+    printf("Building hexagon topology: %d straight + %d parallel nodes\n", 
+           opt.straight_nodes, opt.parallel_nodes);
+    
+    const int parallel_path_length = opt.parallel_nodes; 
+    const int parallel_paths = 2;       
+    
+    int straight_nodes = opt.straight_nodes;
+    int before_split = straight_nodes / 2;
+    int after_join = straight_nodes - before_split;
+    
+    std::vector<hipGraphNode_t> nodes;
+    
+    // Step 1: straight line before split
+    hipGraphNode_t last_before_split = {};
+    for (int i = 0; i < before_split; ++i) {
+      if (i == 0) {
+        last_before_split = add_kernel_node(graph, nullptr, 0, opt.iters);
+      } else {
+        hipGraphNode_t n = add_kernel_node(graph, &last_before_split, 1, opt.iters);
+        last_before_split = n;
+      }
+      nodes.push_back(last_before_split);
+    }
+    
+    // Step 2: parallel paths
+    std::vector<hipGraphNode_t> path_ends(parallel_paths);
+    for (int path = 0; path < parallel_paths; ++path) {
+      hipGraphNode_t prev = last_before_split;
+      
+      for (int i = 0; i < parallel_path_length; ++i) {
+        hipGraphNode_t* deps = (before_split > 0) ? &prev : nullptr;
+        size_t numDeps = (before_split > 0) ? 1 : 0;
+        
+        if (i == 0 && before_split > 0) {
+          deps = &last_before_split;
+          numDeps = 1;
+        }
+        
+        hipGraphNode_t n = add_kernel_node(graph, deps, numDeps, opt.iters);
+        nodes.push_back(n);
+        prev = n;
+      }
+      path_ends[path] = prev;
+    }
+    
+    // Step 3: straight line after join
+    if (after_join > 0) {
+      hipGraphNode_t join_node = add_kernel_node(graph, path_ends.data(), path_ends.size(), opt.iters);
+      nodes.push_back(join_node);
+      
+      hipGraphNode_t prev = join_node;
+      for (int i = 1; i < after_join; ++i) {
+        hipGraphNode_t n = add_kernel_node(graph, &prev, 1, opt.iters);
+        nodes.push_back(n);
+        prev = n;
+      }
+    }
+    
+    printf("Hexagon topology created: %d total nodes\n", (int)nodes.size());
+  }
+
```

Comment:
You should do a pregraph warmup, ie do an operation that creates "stream", Creating a stream is costly operation for us.  do an event record on null stream and "stream" followed by a device syncrhonize.  And keep that under a new cmdline argument called opt.pregraphwarmup.